### PR TITLE
moving GL backend out of GPU

### PR DESCRIPTION
**Pull request by jherico**
This is part 1 of OpenGL 4.5 support.  This PR breaks the GL backend out of the `gpu` library into a `gpu-gl` library.  Some small adjustments have been made

* improve logging, 
* remove legacy (non OpenGL core profile) code paths
* place the GLBackend into a child namespace
* migrate the preprocessor directives used to turn on and off some features out of the `gl` library and into `gpu-gl`

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 259eb497f0a001315cef949113f850c29401c3af available at: https://gist.github.com/a3ded6c7089ed253ad4e6e8852054815<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: fe3f143853c65d778a50500f1e73981a0287ea87 available at: https://gist.github.com/a3ded6c7089ed253ad4e6e8852054815<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3515c51cf5bdcf7234e9ddbab59dab0cd0abf362 available at: https://gist.github.com/a3ded6c7089ed253ad4e6e8852054815<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 68fabf2e6f7707f05be13c44880d0a1649e02f60 available at: https://gist.github.com/a3ded6c7089ed253ad4e6e8852054815<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d432cbcea73c25fe6d36cc9fb35cb3470ae1549c available at: https://gist.github.com/a3ded6c7089ed253ad4e6e8852054815<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6442b022beeb6145d276e0bd6df824e5c2b165ed available at: https://gist.github.com/a3ded6c7089ed253ad4e6e8852054815<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 368c7709ca30e6ab400c75d62e368983686f40c9 available at: https://gist.github.com/a3ded6c7089ed253ad4e6e8852054815<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a9c547ca452bf3d87b8a2775926805a66172786b available at: https://gist.github.com/a3ded6c7089ed253ad4e6e8852054815<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b65db048bf5e79466c481b617a03d8a0ccd34472 available at: https://gist.github.com/a3ded6c7089ed253ad4e6e8852054815<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3a60e2c40a67a60c0b46112b5c763750a7df6316 available at: https://gist.github.com/a3ded6c7089ed253ad4e6e8852054815<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 900334777e638b026729346b2c07a37295bd2e7a available at: https://gist.github.com/a3ded6c7089ed253ad4e6e8852054815<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ae80c168d403c077de14291f9c8fd6c65e2f2896 available at: https://gist.github.com/a3ded6c7089ed253ad4e6e8852054815<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.dmg

**Comment by jherico** [comment]
## Testing

This PR should be completely neutral in regards to application behavior.  It merely re-organizes code and cleans up some dead code paths.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5ba8372c2136b7cf95f2e3a58c26e1d94784cf6b available at: https://gist.github.com/a3ded6c7089ed253ad4e6e8852054815<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.dmg

File: libraries/gpu-gl/src/gpu/gl/GLBackendShared.cpp [hunk 1]
```diff
@@ -411,14 +457,16 @@ GLTexelFormat GLTexelFormat::evalGLTexelFormat(const Element& dstFormat, const E
...
 
         default:
-            qCDebug(gpulogging) << "Unknown combination of texel format";
+            qCDebug(gpugllogging) << "Unknown combination of texel format";
         }
         return texel;
     }
 }
+
+} }
```
Comment (samcake) [inline libraries/gpu-gl/src/gpu/gl/GLBackendShared.cpp]: need a new line

File: libraries/gpu/src/gpu/Texture.cpp [hunk 2]
```diff
@@ -884,3 +884,12 @@ Vec3u Texture::evalMipDimensions(uint16 level) const {
...
 }
+
+std::function<uint32(const gpu::Texture& texture)> TEXTURE_ID_RESOLVER;
+
+uint32 Texture::getHardwareId() const {
+    if (TEXTURE_ID_RESOLVER) {
+        return TEXTURE_ID_RESOLVER(*this);
+    }
+    return 0;
+}
```
Comment (samcake) [inline libraries/gpu/src/gpu/Texture.cpp]: need a new line

File: libraries/gpu/CMakeLists.txt [hunk 3]
```diff
@@ -2,6 +2,3 @@ set(TARGET_NAME gpu)
 AUTOSCRIBE_SHADER_LIB(gpu)
 setup_hifi_library()
 link_hifi_libraries(shared gl)
```
Comment (samcake) [inline libraries/gpu/CMakeLists.txt]: Do we still keep gl there then ?

File: libraries/gpu/src/gpu/Texture.cpp (hunk 2 above)
Comment (jherico) [inline libraries/gpu/src/gpu/Texture.cpp]: fixed

File: libraries/gpu-gl/src/gpu/gl/GLBackendShared.cpp (hunk 1 above)
Comment (jherico) [inline libraries/gpu-gl/src/gpu/gl/GLBackendShared.cpp]: fixed

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0d6b0eef59511ab5ce273596277135217cbd3ff8 available at: https://gist.github.com/a3ded6c7089ed253ad4e6e8852054815<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7902.dmg

**Comment by samcake** [comment]
Sounds good to me
I haven't tried the build though

**Comment by kevinmthomas-carpool** [comment]
## QA Test
**QA Tester**: KevinMThomas
**localhost 'createTestBlocks.js' Result**: **localhost 'createTestBlocks.js'** SUCCESS
**localhost ATP Upload + Edit Result**: **localhost ATP** SUCCESS
**Echo Server Audio**: **Echo Server Audio** SUCCESS
**localhost 'harmonicOscillator.js' Result**: **localhost 'harmonicOscillator.js'** SUCCESS
**localhost 'reverbTest.js' Result**: **localhost 'reverbTest.js'** SUCCESS
**localhost 'playaPerformanceTest.js' Result**: **localhost 'playaPerformanceTest.js'** SUCCESS
**playa 'playaPerformanceTest.js' Result**: **playa 'playaPerformanceTest.js'** SUCCESS
**cellscience 'playaPerformanceTest.js' Result**: **cellscience 'playaPerformanceTest.js'** SUCCESS
**QA Pass**: **QA Pass** SUCCESS
**Additional Test Results + Observations**: Successful test. No issues of any kind noted and behavior normal.

**Domain** |localhost |Playa |cellscience
---|---|---|---
**Render Rate**: |60 |55 |60
**Present Rate**: |60 |60 |60
**Present New Rate**: |60 |54 |60
**CPU Buffer Count**: |1094 |2470 |2441
**GPU Buffer Count**: |198 |575 |659
**CPU Memory Usage**: |32534528 |56053760 |30380032
**GPU Memory Usage**: |10317824 |17940480 |15769600
**CPU Texture Count**: |75 |185 |134
**GPU Texture Count**: |59 |141 |121
**GPU Transfer Count**: |0 |0 |0
**CPU Texture Memory Usage**: |28453960 |152819128 |19016776
**GPU Texture Memory Usage**: |346921038 |514218082 |455953629
**GPU Virtual Memory Usage**: |346921038|514218082|455953629
**Triangle Count**: |71679 |271179 |442493
**Triangle Rate**: |4479937 |16948687 |29499533
**API Drawcall Count**: |16 |117 |142
**Drawcall Count**: |16 |115 |142
**Drawcall Rate**: |1000 |7187 |9466
**State Changes Texture Count**:|30 |94 |94
**Pipeline Count**: |17 |31 |24
**Item Opaques**: |11 |252 |681
**Item Translucents**: |2 |53 |24
**Item Lights**: |0 |40 |1
**Timing Opaques**: |175 |1151 |1556
**Timing Translucents**: |11 |191 |313
**Timing Lighting**: |38 |201 |110
**Render Frame**: |743 |2447 |2844

**Pull request merged**